File: plugin/src/main/java/com/stratio/cassandra/lucene/schema/column/Column.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.schema.column;

import com.google.common.base.Objects;
import com.stratio.cassandra.lucene.IndexException;
import org.apache.cassandra.db.marshal.AbstractType;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.regex.Pattern;

/**
 * A cell of a CQL3 logic {@link Column}, which in most cases is different from a storage engine column.
 *
 * @param <T> The type of the column value.
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public final class Column<T>
{

    /**
     * The map name components separator.
     */
    public static final String MAP_SEPARATOR = "$";

    /**
     * The UDT name components separator.
     */
    public static final String UDT_SEPARATOR = ".";

    public static final String UDT_PATTERN = Pattern.quote(UDT_SEPARATOR);
    public static final String MAP_PATTERN = Pattern.quote(MAP_SEPARATOR);

    private static final Pattern NAME_PATTERN = Pattern.compile("[^(\\$|\\.)]*[\\.[^(\\$|\\.)]]*[\\$[^(\\$|\\.)]]*");

    /**
     * The full qualified name, with UDT and map qualifiers.
     */
    private final String cellName;

    private final List<String> udtNames;
    private final List<String> mapNames;

    /**
     * The column's value as {@link ByteBuffer}.
     */
    private final T composedValue;

    /**
     * The column's value as {@link ByteBuffer}.
     */
    private final ByteBuffer decomposedValue;

    /**
     * The column's Cassandra type.
     */
    private final AbstractType<T> type;

    private final boolean isMultiCell;

    private final int localDeletionTime;

    /**
     * Builds a new {@link Column} with the specified name, name suffix, value, and type.
     *
     * @param cellName        The name of the base cell.
     * @param udtNames        The child UDT fields.
     * @param mapNames        The child map keys.
     * @param decomposedValue The decomposed value of the column to be created.
     * @param composedValue   The composed value of the column to be created.
     * @param type            The type/marshaller of the column to be created.
     * @param isMultiCell     If the column is a multiCell column (not frozen Collections).
     */
    Column(String cellName,
           List<String> udtNames,
           List<String> mapNames,
           ByteBuffer decomposedValue,
           T composedValue,
           AbstractType<T> type,
           boolean isMultiCell,
           int localDeletionTime)
    {
        this.cellName = cellName;
        this.udtNames = udtNames;
        this.mapNames = mapNames;
        this.composedValue = composedValue;
        this.decomposedValue = decomposedValue;
        this.type = type;
        this.isMultiCell = isMultiCell;
        this.localDeletionTime = localDeletionTime;
    }

    public static ColumnBuilder builder(String cellName)
    {
        return new ColumnBuilder(cellName);
    }

    public static boolean isTuple(String name)
    {
        return name.contains(UDT_SEPARATOR);
    }

    public static String check(String name)
    {
        if (!NAME_PATTERN.matcher(name).matches())
        {
            throw new IndexException("Name %s doesn't satisfy the mandatory pattern %s", name, NAME_PATTERN.pattern());
        }
        return name;
    }

    public static String getMapperName(String field)
    {
        return field.split(MAP_PATTERN)[0];
    }

    public static String getCellName(String field)
    {
        return field.split(UDT_PATTERN)[0].split(MAP_PATTERN)[0];
    }

    public String getMapperName()
    {
        return cellName + getUDTSuffix();
    }

    /**
     * Returns the column name.
     *
     * @return the column name.
     */
    public String getCellName()
    {
        return cellName;
    }

    /**
     * Returns the full name, which is formed by the column name and the suffix.
     *
     * @return The full name, which is formed by the column name and the suffix.
     */
    public String getFullName()
    {
        return cellName + getUDTSuffix() + getMapSuffix();
    }

    private String getUDTSuffix()
    {
        String result = "";
        for (String udtName : udtNames)
        {
            result += UDT_SEPARATOR + udtName;
        }
        return result;
    }

    private String getMapSuffix()
    {
        String result = "";
        for (String mapName : mapNames)
        {
            result += MAP_SEPARATOR + mapName;
        }
        return result;
    }

    /**
     * Returns if the column isLive now.
     *
     * @return If the column isLive now.
     */
    public boolean isLive(long now)
    {
        return (int) (now / 1000) < localDeletionTime;
    }

    /**
     * Returns the field column name appending the suffix.
     *
     * @param field A base field name.
     * @return The full column name appending the suffix.
     */
    public String getFieldName(String field)
    {
        return field + getMapSuffix();
    }

    /**
     * Returns the {@link ByteBuffer} serialized value.
     *
     * @return the {@link ByteBuffer} serialized value.
     */
    public ByteBuffer getDecomposedValue()
    {
        return decomposedValue;
    }

    /**
     * Returns the Java column value.
     *
     * @return The Java column value.
     */
    public T getComposedValue()
    {
        return composedValue;
    }

    /**
     * Returns the Cassandra column type.
     *
     * @return The Cassandra column type.
     */
    public AbstractType<T> getType()
    {
        return type;
    }

    /**
     * Returns if this Column is a multiCell column (not frozen Collections).
     *
     * @return if this Column is a multiCell column (not frozen Collections).
     */
    public boolean isMultiCell()
    {
        return isMultiCell;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
        return Objects.toStringHelper(this)
                .add("fullName", getFullName())
                .add("composedValue", getComposedValue())
                .add("type", type.getClass().getSimpleName())
                .toString();
    }
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/core/column/Column.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.core.column

import java.util.regex.Pattern

import com.google.common.base.MoreObjects
import com.stratio.cassandra.lucene.IndexException
import org.apache.commons.lang3.StringUtils

/**
  * A cell of a CQL3 logic column, which in most cases is different from a storage engine column.
  *
  * @param cellName     the name of the base cell
  * @param udtNames     the UDT fields
  * @param mapNames     the map keys
  * @param deletionTime the deletion time in seconds
  * @param value        the optional value
  * @tparam A the value type
  * @author Andres de la Pena `adelapena@stratio.com`
  */
case class Column[A](cellName: String,
                     udtNames: List[String] = Nil,
                     mapNames: List[String] = Nil,
                     deletionTime: Int = Column.NO_DELETION_TIME,
                     value: Option[A] = None) {

  if (StringUtils.isBlank(cellName)) throw new IndexException("Cell name shouldn't be blank")

  private[this] lazy val udtSuffix = udtNames.foldLeft("")((a, n) => a + Column.UDT_SEPARATOR + n)
  private[this] lazy val mapSuffix = mapNames.foldLeft("")((a, n) => a + Column.MAP_SEPARATOR + n)

  /** The columns field name, composed by cell name, UDT names and map names. */
  lazy val fieldName = cellName + udtSuffix + mapSuffix

  /** The columns mapper name, composed by cell name and UDT names, without map names. */
  lazy val mapperName = cellName + udtSuffix

  lazy val mapperNames:List[String] = cellName :: udtNames

  /** Returns a copy of this with the specified name appended to the list of UDT names. */
  def withUDTName(name: String): Column[_] =
    copy(udtNames = udtNames :+ name)

  /** Returns a copy of this with the specified name appended to the list of map names. */
  def withMapName(name: String): Column[_] =
    copy(mapNames = mapNames :+ name)

  /** Returns a copy of this with the specified deletion UNIX time in seconds. */
  def withDeletionTime(deletionTime: Int): Column[_] =
    copy(deletionTime = deletionTime)

  /** Returns a copy of this with the specified value. */
  def withValue[B](value: B): Column[B] =
    copy(value = Option(value))

  /** Returns the name for fields. */
  def fieldName(field: String): String =
    field + mapSuffix

  /** Returns true this is a deletion at the specified UNIX timestamp in seconds, false otherwise. */
  def isDeleted(timeInSec: Int): Boolean =
    value.isEmpty || deletionTime <= timeInSec

  /** Returns a [[Columns]] composed by this and the specified column. */
  def +(column: Column[_]): Columns =
    Columns(this, column)


  /** Returns a [[Columns]] composed by this and the specified columns. */
  def +(columns: Columns): Columns =
    Columns(this) + columns

  override def toString: String =
    MoreObjects.toStringHelper(this)
      .add("cell", cellName)
      .add("name", fieldName)
      .add("value", value.getOrElse("null"))
      .add("deletionTime", deletionTime)
      .toString
}

object Column {

  val NO_DELETION_TIME: Int = Int.MaxValue

  private val UDT_SEPARATOR: String = "."
  private val MAP_SEPARATOR: String = "$"

  private[this] val UDT_PATTERN: String = Pattern.quote(UDT_SEPARATOR)
  private[this] val MAP_PATTERN: String = Pattern.quote(MAP_SEPARATOR)

  def apply(cellName: String): Column[_] =
    new Column(cellName = cellName)

  def parse(name: String): Column[_] = {
    val x = name.split(MAP_PATTERN)
    val mapNames = x.drop(1).toList
    val y = x.head.split(UDT_PATTERN)
    val cellName = y.head
    val udtNames = y.drop(1).toList
    new Column(cellName, udtNames, mapNames)
  }
}
